expression/util.go

```diff
@@ -707,6 +707,165 @@ func pushNotAcrossArgs(ctx sessionctx.Context, exprs []Expression, not bool) ([]
+		for i, cnfItem := range cnfItems {
+			newExpr, curDowncast := eliminateCastFunction(sctx, cnfItem)
+			rmCastItems[i] = newExpr
+			if curDowncast {
+				rmCast = true
+			}
+		}
+		if rmCast {
+			// compose the new CNF expression.
+			return ComposeCNFCondition(sctx, rmCastItems...), true
+		}
+		return expr, false
+	case ast.EQ, ast.NullEQ, ast.LE, ast.GE, ast.LT, ast.GT:
+		// for case: eq(cast(test.t2.a, varchar(100), "aaaaa"), once t2.a is covered by index or pk, try deconstructing it out.
+		if cast, ok := f.GetArgs()[0].(*ScalarFunction); ok && cast.FuncName.L == ast.Cast {
+			// eg: if (cast(A) EQ const) with incompatible collation, even if cast is eliminated, the condition still can not be used to build range.
+			if cast.RetType.EvalType() == types.ETString && !collate.CompatibleCollate(cast.RetType.GetCollate(), collation) {
+				return expr, false
+			}
+			if constVal, ok := f.GetArgs()[1].(*Constant); ok {
+				val, err := constVal.Eval(chunk.Row{})
+				if err != nil || (!sctx.GetSessionVars().RegardNULLAsPoint && val.IsNull()) {
+					// treat col<=>null as range scan instead of point get to avoid incorrect results
+					// when nullable unique index has multiple matches for filter x is null
+					return expr, false
+				}
+				// the direct args of cast function should be column.
+				c, ok := cast.GetArgs()[0].(*Column)
+				if !ok {
+					return expr, false
+				}
+
+				// current only consider varchar and integer
+				if !noPrecisionLossCastCompatible(cast.RetType, c.RetType) {
+					return expr, false
+				}
+
+				// the column is covered by indexes, deconstructing it out.
+				return NewFunctionInternal(sctx, f.FuncName.L, f.RetType, c, f.GetArgs()[1]), true
+			}
+		}
+		// for case: eq("aaaaa"， cast(test.t2.a, varchar(100)), once t2.a is covered by index or pk, try deconstructing it out.
+		if cast, ok := f.GetArgs()[1].(*ScalarFunction); ok && cast.FuncName.L == ast.Cast {
+			// eg: if (cast(A) EQ const) with incompatible collation, even if cast is eliminated, the condition still can not be used to build range.
+			if cast.RetType.EvalType() == types.ETString && !collate.CompatibleCollate(cast.RetType.GetCollate(), collation) {
+				return expr, false
+			}
+			if constVal, ok := f.GetArgs()[0].(*Constant); ok {
+				val, err := constVal.Eval(chunk.Row{})
+				if err != nil || (!sctx.GetSessionVars().RegardNULLAsPoint && val.IsNull()) {
+					// treat col<=>null as range scan instead of point get to avoid incorrect results
+					// when nullable unique index has multiple matches for filter x is null
+					return expr, false
+				}
+				// the direct args of cast function should be column.
+				c, ok := cast.GetArgs()[0].(*Column)
+				if !ok {
+					return expr, false
+				}
+
+				// current only consider varchar and integer
+				if !noPrecisionLossCastCompatible(cast.RetType, c.RetType) {
+					return expr, false
+				}
+				return NewFunctionInternal(sctx, f.FuncName.L, f.RetType, f.GetArgs()[0], c), true
```

Comment:
I think this part is quite similar to L787-L809 above and L845-L864 below. We can extract them to a separate function.